<a href="https://colab.research.google.com/github/lalaggv2/CAD/blob/master/Taller_grupal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## TALLER

In [ ]:
import pandas as pd
import numpy as np

ruta = "https://drive.google.com/uc?id=1PWsoCyTqhopPScpNiidgFUrwTpWQk6T2"
df = pd.read_csv(ruta, index_col="Unnamed: 0")

In [ ]:
df.head()

,symboling,normalized-losses,make,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,length,...,compression-ratio,horsepower,peak-rpm,city-mpg,highway-mpg,price,city-L/100km,horsepower-binned,diesel,gas
0,3,122,alfa-romero,std,two,convertible,rwd,front,88.6,0.811148,...,9.0,111.0,5000.0,21,27,13495.0,11.190476,Medium,0,1
1,3,122,alfa-romero,std,two,convertible,rwd,front,88.6,0.811148,...,9.0,111.0,5000.0,21,27,16500.0,11.190476,Medium,0,1
2,1,122,alfa-romero,std,two,hatchback,rwd,front,94.5,0.822681,...,9.0,154.0,5000.0,19,26,16500.0,12.368421,Medium,0,1
3,2,164,audi,std,four,sedan,fwd,front,99.8,0.848630,...,10.0,102.0,5500.0,24,30,13950.0,9.791667,Medium,0,1
4,2,164,audi,std,four,sedan,4wd,front,99.4,0.848630,...,8.0,115.0,5500.0,18,22,17450.0,13.055556,Medium,0,1


1. ¿Cuántas filas y columnas tiene el dataset?

In [ ]:
print(df.shape)

(201, 29)


In [ ]:
filas, columnas = df.shape
print(f"El dataset tiene {filas} filas y {columnas} columnas.")

El dataset tiene 201 filas y 29 columnas.


2. ¿Qué tipos de datos tiene cada columna?

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 201 entries, 0 to 200
Data columns (total 29 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  201 non-null    int64  
 2   make               201 non-null    object 
 3   aspiration         201 non-null    object 
 4   num-of-doors       201 non-null    object 
 5   body-style         201 non-null    object 
 6   drive-wheels       201 non-null    object 
 7   engine-location    201 non-null    object 
 8   wheel-base         201 non-null    float64
 9   length             201 non-null    float64
 10  width              201 non-null    float64
 11  height             201 non-null    float64
 12  curb-weight        201 non-null    int64  
 13  engine-type        201 non-null    object 
 14  num-of-cylinders   201 non-null    object 
 15  engine-size        201 non-null    int64  
 16  fuel-system        201 non-null

3. ¿Cuántos valores nulos hay por columna?
 ¿Qué columnas tienen más del 10% de nulos?

In [ ]:
nulos = df.isnull().sum()
nulos

,0
symboling,0
normalized-losses,0
make,0
aspiration,0
num-of-doors,0
body-style,0
drive-wheels,0
engine-location,0
wheel-base,0
length,0


In [ ]:
# Cálculo del porcentaje de nulos
porcentaje = (nulos / len(df) * 100)

# Filtro de nulos
muchos_nulos = porcentaje[porcentaje>10]

print(muchos_nulos)
print(f"El dataset no tiene más del 10% de nulos")

Series([], dtype: float64)
El dataset no tiene más del 10% de nulos


4.  El gerente de ventas quiere lanzar una promoción para autos económicos. ¿Cuáles son los 5 autos más baratos del catálogo? Muestra marca, precio y número de puertas.


In [ ]:
# Display para forzar a python a mostrar toda la información del código

display(df["price"])
display(df["make"])
display(df["num-of-doors"])

,price
0,13495.0
1,16500.0
2,16500.0
3,13950.0
4,17450.0
...,...
196,16845.0
197,19045.0
198,21485.0
199,22470.0


,make
0,alfa-romero
1,alfa-romero
2,alfa-romero
3,audi
4,audi
...,...
196,volvo
197,volvo
198,volvo
199,volvo


,num-of-doors
0,two
1,two
2,two
3,four
4,four
...,...
196,four
197,four
198,four
199,four


In [ ]:
carros_baratos = df.sort_values(by="price", ascending=True).head(5)
más_baratos = carros_baratos[["make","price","num-of-doors"]]
print(más_baratos)

           make   price num-of-doors
134      subaru  5118.0          two
17    chevrolet  5151.0          two
47        mazda  5195.0          two
146      toyota  5348.0          two
73   mitsubishi  5389.0          two


5. ¿Cuántos autos cuestan más de $30,000? ¿Qué porcentaje del total representan?

In [ ]:
carros_caros = df[df["price"] > 30000]
display(print(f"Carros cuyo precio supera los 30 mil:\n{carros_caros[["price"]]}"))
porcentaje_carros_caros = (len(carros_caros) / len(df)) * 100
display(print(f"Los carros más caros representan el {porcentaje_carros_caros} %"))

Carros cuyo precio supera los 30 mil:
       price
14   30760.0
15   41315.0
16   36880.0
44   32250.0
45   35550.0
46   36000.0
67   31600.0
68   34184.0
69   35056.0
70   40960.0
71   45400.0
123  32528.0
124  34028.0
125  37028.0


None

Los carros más caros representan el 6.965174129353234 %


None

6. El departamento de marketing quiere autos con buena relación precio-potencia. Crea una nueva columna llamada precio_por_hp (precio / horsepower) y encuentra el auto con el mejor valor (menor precio por caballo).

In [ ]:
# Verificación de tipo de datos: sí son numéricos
print(df["horsepower"].dtype)
print(df["price"].dtype)

float64
float64


In [ ]:
# Creación de columna
df["precio_por_hp"] = df["price"] / df["horsepower"]

# Filtro para el mejor auto precio-potencia
mejor_carro = df.sort_values(by="precio_por_hp",ascending=True).head(1)

print(f"El carro con mejor relación precio-potencia es: \n{mejor_carro[["make","price","horsepower","precio_por_hp"]]}")

El carro con mejor relación precio-potencia es: 
       make   price  horsepower  precio_por_hp
163  toyota  8449.0       116.0      72.836207


7. ¿Cuál es el precio promedio, mínimo y máximo por tipo de carrocería (body-style)? ¿Qué tipo de carrocería tiene el precio promedio más alto?

7.1. Calcular promedio, mínimo y máximo por body-style

In [ ]:
df.groupby('body-style')['price'].agg(['mean', 'min', 'max']).round(2)

,mean,min,max
body-style,,,
convertible,21890.50,11595.0,37028.0
hardtop,22208.50,8249.0,45400.0
hatchback,9957.44,5118.0,22018.0
sedan,14459.76,5499.0,41315.0
wagon,12371.96,6918.0,28248.0


7.2 Encontrar el tipo de carrocería con mayor precio promedio

In [ ]:
df.groupby('body-style')['price'].mean().idxmax()

'hardtop'

8. ¿Qué marcas tienen el precio máximo por encima de $40,000? ¿Cuántos modelos de esas marcas superan ese umbral?

8.1 Marcas con precio máximo > 40,000

In [ ]:
marcas_caras = df.groupby('make')['price'].max()
marcas_mayores_40000 = marcas_caras[marcas_caras > 40000]

print(marcas_mayores_40000)

make
bmw              41315.0
mercedes-benz    45400.0
Name: price, dtype: float64


8.2 ¿Cuántos modelos de esas marcas superan ese precio?

In [ ]:
df_filtrado = df[df['make'].isin(marcas_mayores_40000.index)]
modelos_mayores_40000 = df_filtrado[df_filtrado['price'] > 40000]

conteo = modelos_mayores_40000['make'].value_counts()

print(conteo)

make
mercedes-benz    2
bmw              1
Name: count, dtype: int64


9. Crea una nueva columna price_category que clasifique:

* 'económico' si price < 10,000

* 'estándar' si 10,000 ≤ price ≤ 20,000

* 'premium' si price > 20,000

Luego, cuenta cuántos autos hay en cada categoría.

9.1 Crear la columna price_category

In [ ]:
def clasificar_precio(precio):
    if precio < 10000:
        return 'económico'
    elif 10000 <= precio <= 20000:
        return 'estándar'
    else:
        return 'premium'

df['price_category'] = df['price'].apply(clasificar_precio)

9.2. Contar cuántos autos hay en cada categoría

In [ ]:
df['price_category'].value_counts()

,count
price_category,
económico,98
estándar,78
premium,25


In [ ]:
df[['price', 'price_category']].head()

,price,price_category
0,13495.0,estándar
1,16500.0,estándar
2,16500.0,estándar
3,13950.0,estándar
4,17450.0,estándar


10. ¿Qué marca tiene, en promedio, el mejor consumo en ciudad (city_mpg)? ¿Y en carretera (highway_mpg)?

In [ ]:
mejor_ciudad = df.groupby('make')['city-mpg'].mean().idxmax()
mejor_carretera = df.groupby('make')['highway-mpg'].mean().idxmax()

print("Mejor consumo en ciudad:", mejor_ciudad)
print("Mejor consumo en carretera:", mejor_carretera)

Mejor consumo en ciudad: chevrolet
Mejor consumo en carretera: chevrolet


Ver los promedios por marca

In [ ]:
df.groupby('make')[['city-mpg', 'highway-mpg']].mean().sort_values(by='city-mpg', ascending=False)

,city-mpg,highway-mpg
make,,
chevrolet,41.000000,46.333333
honda,30.384615,35.461538
volkswagen,28.583333,34.916667
plymouth,28.142857,34.142857
dodge,28.000000,34.111111
toyota,27.500000,32.906250
nissan,27.000000,32.944444
subaru,26.333333,30.750000
mazda,25.705882,31.941176


11. Agrupa por num-of-cylinders y calcula el horsepower promedio. ¿Más cilindros implica más potencia?

11.1. Promedio de horsepower por número de cilindros

In [ ]:
#promedio de potencia por cada tipo de motor (4, 6, 8 cilindros, etc.)
df.groupby('num-of-cylinders')['horsepower'].mean().sort_values().round(2)

,horsepower
num-of-cylinders,
three,48.00
four,90.99
two,109.50
five,118.70
six,161.92
eight,169.50
twelve,262.00


11.2 ordenado de mayor a menor

In [ ]:
df.groupby('num-of-cylinders')['horsepower'].mean().sort_values(ascending=False).round(2)

,horsepower
num-of-cylinders,
twelve,262.00
eight,169.50
six,161.92
five,118.70
two,109.50
four,90.99
three,48.00


12. ¿Qué tipo de combustible (diesel vs gas) tiene mayor precio promedio? ¿Cuál es más eficiente en consumo?

12.1. Precio promedio por tipo de combustible:

In [ ]:
df['fuel_type'] = df.apply(lambda x: 'diesel' if x['diesel'] == 1 else 'gas', axis=1)
df.groupby('fuel_type')['price'].mean().round(2)

,price
fuel_type,
diesel,15838.15
gas,12916.41


12.2Eficiencia

In [ ]:
df.groupby('fuel_type')[['city-mpg', 'highway-mpg']].mean().round(2)

,city-mpg,highway-mpg
fuel_type,,
diesel,30.30,34.75
gas,24.61,30.24


Conclusion: Diesel recorre más distancia con menos combustible

 13. Agrupa por num-of-doors (2 o 4) y calcula el horsepower promedio. ¿Los autos de 2 puertas son más potentes en promedio?

13.1 Promedio de horsepower por número de puertas

In [ ]:
df['horsepower'] = pd.to_numeric(df['horsepower'], errors='coerce')

df.groupby('num-of-doors')['horsepower'].mean().sort_values(ascending=False).round(2)

,horsepower
num-of-doors,
two,107.84
four,100.09


14. Usando map, crea una nueva columna cilindros_categoria que clasifique:

* 'bajo' si num-of-cylinders es 'four' o 'three'
* 'medio' si es 'five' o 'six'
* 'alto' si es 'eight' o 'twelve'

14.1 Crear la clasificación con map

In [ ]:
mapa_cilindros = {
    'three': 'bajo',
    'four': 'bajo',
    'five': 'medio',
    'six': 'medio',
    'eight': 'alto',
    'twelve': 'alto'
}

df['cilindros_categoria'] = df['num-of-cylinders'].map(mapa_cilindros)
df[['num-of-cylinders', 'cilindros_categoria']].head()


,num-of-cylinders,cilindros_categoria
0,four,bajo
1,four,bajo
2,six,medio
3,four,bajo
4,five,medio


15. Usando apply, crea una función que clasifique el price en 3 categorías: 'económico' (<10000), 'estándar' (10000-20000), 'premium' (>20000). Aplica esa función al DataFrame.

In [ ]:
def clasificar_precio(precio):
    if precio < 10000:
        return 'económico'
    elif 10000 <= precio <= 20000:
        return 'estándar'
    else:
        return 'premium'
df['price_category'] = df['price'].apply(clasificar_precio)
df[['price', 'price_category']].head()

,price,price_category
0,13495.0,estándar
1,16500.0,estándar
2,16500.0,estándar
3,13950.0,estándar
4,17450.0,estándar


In [ ]:
df['price_category'].value_counts()

,count
price_category,
económico,98
estándar,78
premium,25


16. Usando el método del rango intercuartil (IQR), detecta outliers en la columna horsepower. ¿Cuántos outliers encuentras?

16.1 Asegurar que horsepower sea numérico

In [ ]:
df['horsepower'] = pd.to_numeric(df['horsepower'], errors='coerce')

In [ ]:
Q1 = df['horsepower'].quantile(0.25)
Q3 = df['horsepower'].quantile(0.75)

IQR = Q3 - Q1

limite_inferior = Q1 - 1.5 * IQR
limite_superior = Q3 + 1.5 * IQR

In [ ]:
outliers = df[(df['horsepower'] < limite_inferior) | (df['horsepower'] > limite_superior)]

print(len(outliers))

outliers[['horsepower']]

5


,horsepower
46,262.0
102,200.0
123,207.0
124,207.0
125,207.0


17. ¿Qué marcas (make) tienen los outliers de horsepower más extremos? Muestra los 3 valores más altos y los 3 más bajos que sean outliers.

18. ¿Tiene sentido eliminar estos outliers o podrían ser valores reales? Justifica tu respuesta desde una perspectiva de negocio.